commit 71b807e1a69940beb31fdeb481c2e54509e8bd32
From: Steve Dickson <steved@redhat.com>
Date: Thu, 1 Mar 2018 13:52:47 -0500
Subject: mount.nfs: minorversion setting is being ignored with the -t flag
 mount -t nfs or mount -t nfs4 and setting minorversion
 should set the v4 minor version. This patch adds a few
 checks to make sure the minor version is set.
 
 The patch also translate the minorversion= option
 to vers=4.x in the arguments passed to the kernel
 when the kernel is later then v3.4.
 
 Signed-off-by: Steve Dickson <steved@redhat.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2059197
Bug: https://lore.kernel.org/all/20180227221318.4866-1-steved@redhat.com/
Origin: upstream, https://git.linux-nfs.org/?p=steved/nfs-utils.git;a=commitdiff;h=71b807e1a69940beb31fdeb481c2e54509e8bd32
Last-Update: 2024-04-03

Index: nfs-utils-1.3.4/utils/mount/network.c
===================================================================
--- nfs-utils-1.3.4.orig/utils/mount/network.c	2024-04-03 14:46:36.504220416 +1300
+++ nfs-utils-1.3.4/utils/mount/network.c	2024-04-03 14:46:36.504220416 +1300
@@ -1266,8 +1266,8 @@
 		}
 	}
 
-	if (!found && strcmp(type, "nfs4") == 0)
-		version_val = type + 3;
+	if (!found && strncmp(type, "nfs", 3) == 0)
+		version_val = "4";
 	else if (!found)
 		return 1;
 	else if (i <= 2 ) {
@@ -1299,9 +1299,14 @@
 		if (!(version->minor = strtol(version_val, &cptr, 10)) && cptr == version_val)
 			goto ret_error;
 		version->v_mode = V_SPECIFIC;
-	} else if (version->major > 3 && *cptr == '\0')
-		version->v_mode = V_GENERAL;
-
+	} else if (version->major > 3 && *cptr == '\0') {
+		version_val = po_get(options, "minorversion");
+		if (version_val != NULL) {
+			version->minor = strtol(version_val, &cptr, 10);
+			version->v_mode = V_SPECIFIC;
+		} else 
+			version->v_mode = V_GENERAL;
+	}
 	if (*cptr != '\0')
 		goto ret_error;
 
Index: nfs-utils-1.3.4/utils/mount/stropts.c
===================================================================
--- nfs-utils-1.3.4.orig/utils/mount/stropts.c	2024-04-03 14:46:36.504220416 +1300
+++ nfs-utils-1.3.4/utils/mount/stropts.c	2024-04-03 14:46:36.504220416 +1300
@@ -755,6 +755,21 @@
 			errno = EINVAL;
 			goto out_fail;
 		}
+	} else if (po_get(options, "minorversion") &&
+		linux_version_code() > MAKE_VERSION(3, 4, 0)) {
+		/*
+	 	 * convert minorversion= into vers=4.x
+	 	 */
+		po_remove_all(options, "minorversion");
+
+		snprintf(version_opt, sizeof(version_opt) - 1,
+			"vers=%lu.%lu", mi->version.major,
+			mi->version.minor);
+
+		if (po_append(options, version_opt) == PO_FAILED) {
+			errno = EINVAL;
+			goto out_fail;
+		}
 	}
 
 	if (!nfs_append_addr_option(sap, salen, options)) {
